### PR TITLE
Kustomize proposal for cross-cutting PodSpec patches

### PR DESCRIPTION
**Pull request by pwittrock**

### CODE REVIEW

File: keps/sig-cli/kustomize-pod-spec-patch.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+type Kustomization struct {
+	// ...
+	
+	patchesPodSpec []PatchPodSpec `json:"patchesPodSpec"`
+	// ...
+}
+
+// ResourceSelector selects Resources that it matches
+type PatchPodSpec struct {
```
Comment (Liujingfang1) [inline keps/sig-cli/kustomize-pod-spec-patch.md]: So `patchesPodSpec` is a kind of patch that can be applied to multiple targets. In `patchesStrategicMerge` and `patchesJson6902`, kustomize uses group, version, kind and name to find the target resource to patch. I think we can also add group, version, kind and name to `patchesPodSpec`. Those fields are all optional and the name could support regular expression.

File: keps/sig-cli/kustomize-pod-spec-patch.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	// matchAnnotations is a map of {key,value} pairs.  A Resource matches if it has
+	// *all* of the annotations in matchAnnotations appear in the Resource metaData.
+	// Null and empty values are not allowed.
+	// +optional
```
Comment (Liujingfang1) [inline keps/sig-cli/kustomize-pod-spec-patch.md]: `LabelSelectors` as well?

File: keps/sig-cli/kustomize-pod-spec-patch.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+```
+
+### Risks and Mitigations
+
+
+## Graduation Criteria
+
+Use customer feedback to determine if we should support:
+
+- `LabelSelectors`
```
Comment (Liujingfang1) [inline keps/sig-cli/kustomize-pod-spec-patch.md]: As I remember, there were Kustomize issues asking for applying patches to resources by matching labels.

File: keps/sig-cli/kustomize-pod-spec-patch.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+## Summary
+
+A common need for users is to inject sidecar containers across many different workloads.
+While overlays and patches allow injection on a per-Resource basis, they don't
+provide a mechanism for injecting the same patch across multiple workloads.
+
+## Motivation
+
+Cross-cutting PodSpec customizations are necessary for usecases such as:
```
Comment (Liujingfang1) [inline keps/sig-cli/kustomize-pod-spec-patch.md]: If current `patchesJson6902` or `patchesStrategicMerge` is extended to patch multiple targets (by regex in names or label selector), can we achieve the cross-cutting PodSpec customizations by them? If not, what is the missing piece?

File: keps/sig-cli/kustomize-pod-spec-patch.md (hunk 1 above)
Comment (pwittrock) [reply keps/sig-cli/kustomize-pod-spec-patch.md]: What is the use case for specifying those?

File: keps/sig-cli/kustomize-pod-spec-patch.md (hunk 3 above)
Comment (pwittrock) [reply keps/sig-cli/kustomize-pod-spec-patch.md]: This is pretty similar.  How much of that was for PodTemplate vs other things?  PodTemplate and Metadata would probably cover everything.

File: keps/sig-cli/kustomize-pod-spec-patch.md (hunk 4 above)
Comment (pwittrock) [reply keps/sig-cli/kustomize-pod-spec-patch.md]: That SGTM.  I think that would be useful for stuff like ObjectMeta as well, which could be cross-cutting.  Would you be ok with me closing this KEP, can you could open up a new one proposing that solution?

**Comment by pwittrock** [comment]
I've seen this issue come up with users needing it 4-5 times, so I think it is a pretty high-value thing to add.

File: keps/sig-cli/kustomize-pod-spec-patch.md (hunk 1 above)
Comment (Liujingfang1) [reply keps/sig-cli/kustomize-pod-spec-patch.md]: For example, applying the patch to all Deployments or applying the patch to all resources with names "app1-*".

File: keps/sig-cli/kustomize-pod-spec-patch.md (hunk 4 above)
Comment (Liujingfang1) [reply keps/sig-cli/kustomize-pod-spec-patch.md]: That sounds good to me. I can open a new one.

**Comment by pwittrock** [comment]
Closing in favor of @Liujingfang1's implementation

**Pull request closed without merging**